Implement motion import

luisa-beerboom opened the pull request:
Closes #1713 

Uses OpenSlides/openslides-meta#14

luisa-beerboom commented:
Up until now the `json_upload` seems to work in all cases involving only the required fields (including reason, which is sometimes required), as well as the number field.

Submitter generation in case of omission seems to be working as well.

The categories have been tested in so far that the `json_upload` seems to work for most common cases (there are still some untested edge cases though)

There are some remaining TODOs and all other fields have not yet been tested at all.
The import too is neither finished, nor tested.

@r-peschke If you wish to have a look at and give feedback to the finished parts sometime next week, please feel free to do so, I'd appreciate that

r-peschke commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,723 @@
+        if (category_name := entry.get("category_name")) and isinstance(
+            category_name, str
+        ):
+            category_prefix = entry.get("category_prefix") or None
```
You don't need   the `or None`. The get-method returns the value, if the dictionary has the attribute or it returns None.

luisa-beerboom replied:
I need `or None` because I need this to ignore empty strings

r-peschke replied:
Then please write a test for this edge case:-).

luisa-beerboom replied:
Already exists

r-peschke replied:
If the `and None` is necessary, there should be a reason why it is necessary. I would expect removing it, changes the behaviour. And this change should be manifested in a failing test

luisa-beerboom replied:
I removed the or None and it still works, seems it was unnecessary

r-peschke commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,723 @@
+        self.row_state = ImportState.ERROR
+        return entry
+
+    def _get_motion_mock_id(self) -> int:
```
Please remove the method, an id is not needed

r-peschke commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,723 @@
+                [*mapped_fields, "id", fieldname],
+                lock_result=False,
+            )
+            for date_id in data:
```
Replace the line by `for date in data.values():`

luisa-beerboom commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,723 @@
+        mapped_fields: List[str] = [],
+        and_filters: List[Filter] = [],
+    ) -> Dict[str, List[Dict[str, Any]]]:
+        lookup: Dict[str, List[Dict[str, Any]]] = {}
```
There is nothing above

r-peschke replied:
> There is nothing above

Sorry, I meant **below**

luisa-beerboom replied:
At what line of what file is that supposed to be? I can't find it

luisa-beerboom replied:
?

luisa-beerboom commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,723 @@
+    def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
+        # transform instance into a correct create/update payload
+        # try to find a pre-existing motion with the same number
+        # if there is one, validate for a motion.create, otherwise for a motion.update
```
Seems like I wrote that the wrong way around, I'll change it

r-peschke reviewed (changes requested):
Some small things.

r-peschke commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,708 @@
+                            )
+                        else:
+                            username_set.add(user)
+                            found_users = lookup.get(user, [])
```
You cannot get more than one user, the username is unique. If you find duplicate users, you should throw an exception and finish the json_upload immediately,

luisa-beerboom replied:
Done

luisa-beerboom commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,708 @@
+                            entry_list.append(
+                                {"value": user, "info": ImportState.WARNING}
+                            )
+                            message_set.add(
```
Done for users

luisa-beerboom replied:
Done for tags

luisa-beerboom replied:
Done in import as well

luisa-beerboom commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,709 @@
+    _operator_username: Optional[str] = None
+    _previous_numbers: List[str]
+
+    _last_motion_mock_id: Optional[int] = None
```
Done

r-peschke commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,712 @@
+                )
+            ),
+            "username",
+            ["meeting_user_ids"],
```
add the field `meeting_ids` to the mapped_fields. This field is the calculated field for membership in a meeting. You don't need to look for the group-membership and consequently this lookup is everything you need concerning the submitter- and supporter-users.

luisa-beerboom replied:
Done in json_upload

luisa-beerboom replied:
Done in import as well

r-peschke commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,712 @@
+            "username",
+            ["meeting_user_ids"],
+        )
+        all_user_ids = [
```
All this code beginning here down to line 609 is not needed, see remark above

luisa-beerboom commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,708 @@
+        has_submitter_error: bool = False
+        for fieldname in ["submitter", "supporter"]:
+            if users := entry.get(f"{fieldname}s_username"):
+                if not isinstance(users, list):
```
I already have tests that check if it works when the submitters/supporters are not in a list.
It works, even after removing the if clauses. I surmise that the if clauses were unnecessary.
Does this address your concerns for this remark?

r-peschke commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,712 @@
+                        err_message, err_users = message_map.get(
+                            "duplicate",
+                            (
+                                f"At least one {fieldname} has been referenced multiple times: ",
```
It's better to gather the `username` in a list,  together with the `fieldname` and create the message one time at the end of the `fieldnames` loop

luisa-beerboom commented:
TODO:

see  #2062

luisa-beerboom commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,708 @@
+                entry_list: list[Dict[str, Any]] = []
+                message_set = set()
+                lookup = (
+                    self.submitter_lookup
```
Done in import as well

luisa-beerboom commented:
Rewrote json_upload tests, found some things about the main class that I thought needed changing/reconsidering, so there are currently some TODOs in the tests and some of them fail (as I've already rewritten them).

Also I am planning to move many of these tests to a separate class for the purpose of using their results in the import test that I've yet to rewrite, so no need to worry about that yet.

@r-peschke would you take a look at what I have so far?

r-peschke commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,607 @@
+            )
+        return self._first_state_id
+
+    def _get_field_array(self, entry: dict[str, Any], fieldname: str) -> list[str]:
```
Seems, that this method is never used

r-peschke commented on openslides_backend/action/actions/motion/json_upload.py:
```diff
@@ -0,0 +1,607 @@
+        fieldname = ""
+        match err["type"]:
+            case MotionErrorType.UNIQUE_NUMBER:
+                fieldname = "number"
```
Could you write a test for this case?

r-peschke reviewed (changes requested):
Sorry, but there are some things.
Tomorrow I'll continue with import, ...

r-peschke commented on openslides_backend/action/actions/motion/import_.py:
```diff
@@ -0,0 +1,644 @@
+                        payload, meeting_id
+                    )
+                for err in errors:
+                    fieldname = ""
```
Maybe the problem will be resolved by other tests, but  there is no test with errors passing this code

r-peschke commented on openslides_backend/action/actions/motion/import_.py:
```diff
@@ -0,0 +1,644 @@
+                        f"Error: Number '{number}' found in different id ({id_} instead of {entry['id']})"
+                    )
+            elif check_result == ResultType.FOUND_MORE_IDS:
+                row["state"] = ImportState.ERROR
```
write tests to provoke this error. There are different combinations possible, i.e 2 on database (technically possible), 2 in payload

r-peschke commented on openslides_backend/action/actions/motion/import_.py:
```diff
@@ -0,0 +1,644 @@
+            if check_result == ResultType.FOUND_ID and id_ != 0:
+                if row["state"] != ImportState.DONE:
+                    row["messages"].append(
+                        f"Error: Row state expected to be '{ImportState.DONE}', but it is '{row['state']}'."
```
Missing test for this error

r-peschke commented on openslides_backend/action/actions/motion/import_.py:
```diff
@@ -0,0 +1,644 @@
+                        f"Invalid JsonUpload data: A data row with state '{ImportState.DONE}' must have an 'id'"
+                    )
+                elif entry["id"] != id_:
+                    row["state"] = ImportState.ERROR
```
Missing test for this error

r-peschke commented on openslides_backend/action/actions/motion/import_.py:
```diff
@@ -0,0 +1,644 @@
+                        not_found.append(tag)
+            if len(different):
+                row["messages"].append(
+                    "Error: Tag search didn't deliver the same result as in the preview: "
```
missing test for this error

r-peschke commented on tests/system/action/motion/test_create.py:
```diff
@@ -148,6 +148,43 @@ def test_create_workflow_id(self) -> None:
+            },
+        )
+        self.assert_status_code(response, 200)
+        motion = self.get_model("motion/1")
```
Here you should use the self.assert_model_exists-method

r-peschke commented on openslides_backend/action/actions/motion/update.py:
```diff
@@ -93,34 +98,17 @@ def prefetch(self, action_data: ActionData) -> None:
+        error_messages = self.get_update_payload_integrity_error_message(
+            instance, motion["meeting_id"]
+        )
+        if len(error_messages):
```
Then there is no need to continue after first error

r-peschke replied:
closed
